libclc/utils/libclc-remangler/LibclcRemangler.cpp

```diff
@@ -147,340 +123,602 @@ class BumpPointerAllocator {
+      if (NestedNamesQTMap.count(N) == 0) {
+        assert(KNN->DataStr.rfind("__spv::", 0) == 0 &&
+               "Unexpected nested prefix");
+        SourceLocation const SL{};
+        RecordDecl *RD = nullptr;
+        if (!SpvNamespace)
+          SpvNamespace = NamespaceDecl::Create(
+              *ContextAST, ContextAST->getTranslationUnitDecl(), false, SL, SL,
+              &ContextAST->Idents.get("__spv", tok::TokenKind::identifier),
+              nullptr);
+        std::string StructName{KNN->DataStr};
+        StructName.erase(0, StructName.find_first_not_of("__spv::"));
+        auto *II =
+            &ContextAST->Idents.get(StructName, tok::TokenKind::identifier);
+        RD = RecordDecl::Create(*ContextAST, TTK_Struct, SpvNamespace, SL, SL,
+                                II);
+        auto *NNS =
+            NestedNameSpecifier::Create(*ContextAST, nullptr, SpvNamespace);
+        auto RecordQT = ContextAST->getRecordType(RD);
+        NNS = NestedNameSpecifier::Create(*ContextAST, NNS, false,
+                                          RecordQT.getTypePtr());
+        auto &EnumName =
+            ContextAST->Idents.get(Res.getBaseTypeIdentifier()->getName());
+        // We need to recreate the enum, now that we have access to all the
+        // namespace/class info.
+        auto *ED = EnumDecl::Create(*ContextAST, RD, SourceLocation(),
+                                    SourceLocation(), &EnumName, nullptr, false,
+                                    false, true);
+        Res = ContextAST->getEnumType(ED);
+        Res = ContextAST->getElaboratedType(ETK_None, NNS, Res);
+        // Store the elaborated type for reuse, this is important as clang uses
+        // substitutions for ET based on the object not the name enclosed in.
+        NestedNamesQTMap[N] = Res;
+      } else
+        Res = NestedNamesQTMap[N];
+    }
 
-    if (encoding->getReturnType())
-      remangleOpenCLCType(encoding->getReturnType(), OB);
+    // Iterate in reversed order to preserve the semantics.
+    for (auto I = PossibleKinds.rbegin(); I != PossibleKinds.rend(); ++I) {
+      switch (I->K) {
+      case Node::Kind::KPointerType: {
+        Res = ContextAST->getPointerType(Res);
+        break;
+      }
+      case Node::Kind::KVectorType: {
+        Res = ContextAST->getVectorType(
+            Res, I->Data, clang::VectorType::VectorKind::GenericVector);
+        break;
+      }
+      case Node::Kind::KQualType: {
+        auto Quals = clang::Qualifiers::fromFastMask(I->Data);
+        Res = ContextAST->getQualifiedType(Res, Quals);
+        break;
+      }
+      case Node::Kind::KVendorExtQualType: {
+        auto AS = getLangASFromTargetAS(I->Data);
+        Res = ContextAST->getAddrSpaceQualType(Res, AS);
+        break;
+      }
+      case Node::Kind::KNestedName: {
+        // Handled already.
+        break;
+      }
+      default: {
+        assert(false && "Unexpected Node Kind.");

```

Comment:
```suggestion
        llvm_unreachable("Unexpected Node Kind.");
```
Just to keep consistent. 😄